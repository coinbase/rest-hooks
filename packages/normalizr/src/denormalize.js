import * as ImmutableUtils from './schemas/ImmutableUtils';
import * as ArrayUtils from './schemas/Array';
import * as ObjectUtils from './schemas/Object';

const unvisitEntity = (id, schema, unvisit, getEntity, cache) => {
  const entity = getEntity(id, schema);
  if (typeof entity !== 'object' || entity === null) {
    return [entity, false];
  }

  if (!cache[schema.key]) {
    cache[schema.key] = {};
  }

  let found = true;
  if (!cache[schema.key][id]) {
    // Ensure we don't mutate it non-immutable objects
    const entityCopy = ImmutableUtils.isImmutable(entity)
      ? entity
      : { ...entity };

    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    cache[schema.key][id] = entityCopy;
    [cache[schema.key][id], found] = schema.denormalize(entityCopy, unvisit);
  }

  return [cache[schema.key][id], found];
};

const getUnvisit = entities => {
  const cache = {};
  const getEntity = getEntities(entities);

  return [
    function unvisit(input, schema) {
      if (!schema) return [input, true];

      if (
        typeof schema === 'object' &&
        (!schema.denormalize || typeof schema.denormalize !== 'function')
      ) {
        const method = Array.isArray(schema)
          ? ArrayUtils.denormalize
          : ObjectUtils.denormalize;
        return method(schema, input, unvisit);
      }

      // null is considered intentional, thus always 'found' as true
      if (input === null) {
        return [input, true];
      }

      if (
        typeof schema.getId === 'function' &&
        typeof schema.normalize === 'function'
      ) {
        // unvisitEntity just can't handle undefined
        if (input === undefined) {
          return [input, false];
        }
        return unvisitEntity(input, schema, unvisit, getEntity, cache);
      }

      if (typeof schema.denormalize === 'function') {
        return schema.denormalize(input, unvisit);
      }

      return [input, true];
    },
    cache,
  ];
};

const getEntities = entities => {
  const isImmutable = ImmutableUtils.isImmutable(entities);

  return (entityOrId, schema) => {
    const schemaKey = schema.key;

    if (typeof entityOrId === 'object') {
      return entityOrId;
    }

    if (isImmutable) {
      return entities.getIn([schemaKey, entityOrId.toString()]);
    }

    return entities[schemaKey] && entities[schemaKey][entityOrId];
  };
};

export const denormalize = (input, schema, entities) => {
  /* istanbul ignore next */
  // eslint-disable-next-line no-undef
  if (process.env.NODE_ENV !== 'production' && schema === undefined)
    throw new Error('shema needed');
  if (typeof input !== 'undefined') {
    const [unvisit, cache] = getUnvisit(entities);
    return [...unvisit(input, schema), cache];
  }
  return [undefined, false, {}];
};

export const denormalizeSimple = (input, schema, entities) =>
  denormalize(input, schema, entities).slice(0, 2);
