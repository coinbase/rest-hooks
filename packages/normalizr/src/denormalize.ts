import { isImmutable } from './schemas/ImmutableUtils';
import { denormalize as arrayDenormalize } from './schemas/Array';
import { denormalize as objectDenormalize } from './schemas/Object';
import {
  Denormalize,
  DenormalizeNullable,
  Schema,
  DenormalizeCache,
  UnvisitFunction,
} from './types';
import Entity, { isEntity } from './entities/Entity';
import { DELETED } from './special';
import { EntityInterface } from './schema';
import WeakListMap from './WeakListMap';

const unvisitEntity = (
  id: any,
  schema: any,
  unvisit: UnvisitFunction,
  getEntity: (
    entityOrId: Record<string, any> | string,
    schema: typeof Entity,
  ) => EntityInterface | typeof DELETED,
  localCache: Record<string, Record<string, any>>,
  entityCache: DenormalizeCache['entities'],
): [
  denormalized: EntityInterface | undefined,
  found: boolean,
  deleted: boolean,
] => {
  const entity = getEntity(id, schema);
  if (entity === DELETED) {
    return [undefined, true, true];
  }
  if (typeof entity !== 'object' || entity === null) {
    return [entity, false, false];
  }

  if (localCache[schema.key] === undefined) {
    localCache[schema.key] = {};
  }

  let found = true;
  let deleted = false;
  if (!localCache[schema.key][id]) {
    const globalKey: EntityInterface[] = [entity];
    const wrappedUnvisit = withTrackedEntities(unvisit, globalKey);

    if (!entityCache[schema.key]) entityCache[schema.key] = {};
    if (!entityCache[schema.key][id])
      entityCache[schema.key][id] = new WeakListMap();
    const globalCacheEntry = entityCache[schema.key][id];

    const entityCopy = isImmutable(entity) ? entity : schema.fromJS(entity);
    // Need to set this first so that if it is referenced further within the
    // denormalization the reference will already exist.
    localCache[schema.key][id] = entityCopy;
    [localCache[schema.key][id], found, deleted] = schema.denormalize(
      entityCopy,
      wrappedUnvisit,
    );

    if (!globalCacheEntry.has(globalKey)) {
      globalCacheEntry.set(globalKey, localCache[schema.key][id]);
    } else {
      // localCache is only used before this point for recursive relationships
      // since recursive relationships must all referentially change if *any* do, we either
      // get the correct one here, or will never find the same version in the cache
      localCache[schema.key][id] = globalCacheEntry.get(globalKey);
    }
  }
  return [localCache[schema.key][id], found, deleted];
};

const getUnvisit = (
  entities: Record<string, Record<string, any>>,
  entityCache: DenormalizeCache['entities'],
  resultCache: WeakListMap<object, any>,
  localCache: Record<string, Record<string, any>>,
) => {
  const getEntity = getEntities(entities);

  function unvisit(
    input: any,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] {
    if (!schema) return [input, true, false];

    if (!schema.denormalize || typeof schema.denormalize !== 'function') {
      if (typeof schema === 'function') {
        if (input instanceof schema) return [input, true, false];
        return [new schema(input), true, false];
      } else if (typeof schema === 'object') {
        const method = Array.isArray(schema)
          ? arrayDenormalize
          : objectDenormalize;
        return method(schema, input, wrappedUnvisit);
      }
    }

    // null is considered intentional, thus always 'found' as true
    if (input === null) {
      return [input, true, false];
    }

    if (isEntity(schema)) {
      // unvisitEntity just can't handle undefined
      if (input === undefined) {
        return [input, false, false];
      }
      return unvisitEntity(
        input,
        schema,
        wrappedUnvisit,
        getEntity,
        localCache,
        entityCache,
      );
    }

    if (typeof schema.denormalize === 'function') {
      return schema.denormalize(input, wrappedUnvisit);
    }

    return [input, true, false];
  }

  const globalKey: EntityInterface[] = [];
  const wrappedUnvisit = withTrackedEntities(unvisit, globalKey);

  return (
    input: any,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] => {
    globalKey.push(input);
    const ret = unvisit(input, schema);
    // in the case where WeakMap cannot be used
    // this test ensures null is properly excluded from WeakMap
    if (Object(input) !== input) return ret;

    if (!resultCache.has(globalKey)) {
      resultCache.set(globalKey, ret[0]);
      return ret;
    } else {
      return [resultCache.get(globalKey), ret[1], ret[2]];
    }
  };
};

const getEntities = (entities: Record<string, any>) => {
  const entityIsImmutable = isImmutable(entities);

  return (entityOrId: Record<string, any> | string, schema: typeof Entity) => {
    const schemaKey = schema.key;

    if (typeof entityOrId === 'object') {
      return entityOrId;
    }

    if (entityIsImmutable) {
      return entities.getIn([schemaKey, entityOrId]);
    }

    return entities[schemaKey] && entities[schemaKey][entityOrId];
  };
};

type DenormalizeReturn<S extends Schema> =
  | [
      denormalized: Denormalize<S>,
      found: true,
      deleted: false,
      resolvedEntities: Record<string, Record<string, any>>,
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: boolean,
      deleted: true,
      resolvedEntities: Record<string, Record<string, any>>,
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: false,
      deleted: boolean,
      resolvedEntities: Record<string, Record<string, any>>,
    ];

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const denormalize = <S extends Schema>(
  input: unknown,
  schema: S | undefined,
  entities: any,
  entityCache: DenormalizeCache['entities'] = {},
  resultCache: WeakListMap<object, any> = new WeakListMap(),
): DenormalizeReturn<S> => {
  // undefined mean don't do anything
  if (schema === undefined) {
    return [input, true, false, {}] as [any, boolean, boolean, any];
  }
  if (input === undefined) {
    return [undefined, false, false, {}] as [any, boolean, boolean, any];
  }
  const resolvedEntities: Record<string, Record<string, any>> = {};
  const unvisit = getUnvisit(
    entities,
    entityCache,
    resultCache,
    resolvedEntities,
  );
  return [...unvisit(input, schema), resolvedEntities] as [
    any,
    boolean,
    boolean,
    any,
  ];
};

export const denormalizeSimple = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
  entityCache: DenormalizeCache['entities'] = {},
  resultCache: WeakListMap<object, any> = new WeakListMap(),
):
  | [denormalized: Denormalize<S>, found: true, deleted: false]
  | [denormalized: DenormalizeNullable<S>, found: boolean, deleted: true]
  | [denormalized: DenormalizeNullable<S>, found: false, deleted: boolean] =>
  denormalize(input, schema, entities, entityCache, resultCache).slice(
    0,
    3,
  ) as any;

function withTrackedEntities(
  unvisit: UnvisitFunction,
  globalKey: EntityInterface<any>[],
) {
  // every time we nest, we want to unwrap back to the top.
  // this is due to only needed the next level of nested entities for lookup
  const originalUnvisit = unvisit.og || unvisit;
  const wrappedUnvisit = (input: any, schema: any) => {
    const ret: [any, boolean, boolean] = originalUnvisit(input, schema);
    // pass over undefined in key
    if (ret[0] && schema && isEntity(schema)) {
      /* istanbul ignore else */
      if (Object(ret[0]) === ret[0]) {
        globalKey.push(ret[0]);
      } else if (process.env.NODE_ENV !== 'production') {
        throw new Error(
          `Unexpected primitive found during denormalization\nFound: ${ret[0]}\nExpected entity: ${schema}`,
        );
      }
    }

    return ret;
  };
  wrappedUnvisit.og = unvisit;
  return wrappedUnvisit;
}
