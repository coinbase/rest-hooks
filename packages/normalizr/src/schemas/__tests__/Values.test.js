// eslint-env jest
import { fromJS } from 'immutable';

import { denormalizeSimple as denormalize } from '../../denormalize';
import { normalize, schema } from '../../';
import IDEntity from '../../entities/IDEntity';

describe(`${schema.Values.name} normalization`, () => {
  test('normalizes the values of an object with the given schema', () => {
    class Cat extends IDEntity {}
    class Dog extends IDEntity {}
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      (entity, key) => entity.type,
    );

    expect(
      normalize(
        {
          fido: { id: '1', type: 'dogs' },
          fluffy: { id: '1', type: 'cats' },
        },
        valuesSchema,
      ),
    ).toMatchSnapshot();
  });

  test('can use a function to determine the schema when normalizing', () => {
    class Cat extends IDEntity {}
    class Dog extends IDEntity {}
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      (entity, key) => `${entity.type}s`,
    );

    expect(
      normalize(
        {
          fido: { id: '1', type: 'dog' },
          fluffy: { id: '1', type: 'cat' },
          jim: { id: '2', type: 'lizard' },
        },
        valuesSchema,
      ),
    ).toMatchSnapshot();
  });

  test('filters out null and undefined values', () => {
    class Cat extends IDEntity {}
    class Dog extends IDEntity {}
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      (entity, key) => entity.type,
    );

    expect(
      normalize(
        {
          fido: undefined,
          milo: null,
          fluffy: { id: '1', type: 'cats' },
        },
        valuesSchema,
      ),
    ).toMatchSnapshot();
  });
});

describe(`${schema.Values.name} denormalization`, () => {
  test('denormalizes the values of an object with the given schema', () => {
    class Cat extends IDEntity {}
    class Dog extends IDEntity {}
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      (entity, key) => entity.type,
    );

    const entities = {
      Cat: { '1': { id: '1', type: 'cats' } },
      Dog: { '1': { id: '1', type: 'dogs' } },
    };

    expect(
      denormalize(
        {
          fido: { id: '1', schema: 'dogs' },
          fluffy: { id: '1', schema: 'cats' },
        },
        valuesSchema,
        entities,
      ),
    ).toMatchSnapshot();

    expect(
      denormalize(
        {
          fido: { id: '1', schema: 'dogs' },
          fluffy: { id: '1', schema: 'cats' },
        },
        valuesSchema,
        fromJS(entities),
      ),
    ).toMatchSnapshot();
  });

  test('denormalizes with missing entity should have false second value', () => {
    class Cat extends IDEntity {}
    class Dog extends IDEntity {}
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      (entity, key) => entity.type,
    );

    const entities = {
      Cat: { '1': { id: '1', type: 'cats' } },
      Dog: { '1': { id: '1', type: 'dogs' } },
    };

    expect(
      denormalize(
        {
          fido: { id: '1', schema: 'dogs' },
          fluffy: { id: '1', schema: 'cats' },
          prancy: { id: '5', schema: 'cats' },
        },
        valuesSchema,
        entities,
      ),
    ).toMatchSnapshot();

    expect(
      denormalize(
        {
          fido: { id: '1', schema: 'dogs' },
          fluffy: { id: '1', schema: 'cats' },
          prancy: { id: '5', schema: 'cats' },
        },
        valuesSchema,
        fromJS(entities),
      ),
    ).toMatchSnapshot();
  });
});
