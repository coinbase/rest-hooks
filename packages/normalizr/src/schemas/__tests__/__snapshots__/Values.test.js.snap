// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValuesSchema denormalization denormalizes the values of an object with the given schema 1`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  true,
  true,
]
`;

exports[`ValuesSchema denormalization denormalizes the values of an object with the given schema 2`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  true,
  true,
]
`;

exports[`ValuesSchema denormalization denormalizes with deleted entity should have false third value 1`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  true,
  false,
]
`;

exports[`ValuesSchema denormalization denormalizes with deleted entity should have false third value 2`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  true,
  false,
]
`;

exports[`ValuesSchema denormalization denormalizes with missing entity should have false second value 1`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  false,
  true,
]
`;

exports[`ValuesSchema denormalization denormalizes with missing entity should have false second value 2`] = `
Array [
  Object {
    "fido": Dog {
      "id": "1",
      "type": "dogs",
    },
    "fluffy": Cat {
      "id": "1",
      "type": "cats",
    },
  },
  false,
  true,
]
`;

exports[`ValuesSchema denormalization works on complex object 1`] = `
Object {
  "data": Object {
    "estimates": Object {
      "BTC": Estimate {
        "coinbase_fees": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": Object {
          "crypto": "BTC",
          "local": "USD",
          "rate": "6820.07",
        },
        "exchange_to_proceeds": Object {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "fee_per_kb": Object {
          "amount": "0.00016566",
          "currency": "BTC",
        },
        "min_order_size": Object {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": Object {
          "amount": "0.00054147",
          "currency": "BTC",
        },
      },
      "ETH": Estimate {
        "coinbase_fees": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": Object {
          "crypto": "ETH",
          "local": "USD",
          "rate": "197.07",
        },
        "exchange_to_proceeds": Object {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": Object {
          "amount": "0.03795",
          "currency": "ETH",
        },
        "fee_per_kb": Object {
          "amount": "0.00086",
          "currency": "ETH",
        },
        "min_order_size": Object {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": Object {
          "amount": "2.53",
          "currency": "ETH",
        },
      },
    },
  },
}
`;

exports[`ValuesSchema normalization can use a function to determine the schema when normalizing 1`] = `
Object {
  "entities": Object {
    "Cat": Object {
      "1": Cat {
        "id": "1",
        "type": "cat",
      },
    },
    "Dog": Object {
      "1": Dog {
        "id": "1",
        "type": "dog",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "fido": Object {
      "id": "1",
      "schema": "dogs",
    },
    "fluffy": Object {
      "id": "1",
      "schema": "cats",
    },
    "jim": Object {
      "id": "2",
      "type": "lizard",
    },
  },
}
`;

exports[`ValuesSchema normalization filters out null and undefined values 1`] = `
Object {
  "entities": Object {
    "Cat": Object {
      "1": Cat {
        "id": "1",
        "type": "cats",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "fluffy": Object {
      "id": "1",
      "schema": "cats",
    },
  },
}
`;

exports[`ValuesSchema normalization normalizes the values of an object with the given schema 1`] = `
Object {
  "entities": Object {
    "Cat": Object {
      "1": Cat {
        "id": "1",
        "type": "cats",
      },
    },
    "Dog": Object {
      "1": Dog {
        "id": "1",
        "type": "dogs",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "fido": Object {
      "id": "1",
      "schema": "dogs",
    },
    "fluffy": Object {
      "id": "1",
      "schema": "cats",
    },
  },
}
`;

exports[`ValuesSchema normalization works on complex object 1`] = `
Object {
  "entities": Object {
    "Estimate": Object {
      "BTC": Estimate {
        "coinbase_fees": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": Object {
          "crypto": "BTC",
          "local": "USD",
          "rate": "6820.07",
        },
        "exchange_to_proceeds": Object {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "fee_per_kb": Object {
          "amount": "0.00016566",
          "currency": "BTC",
        },
        "min_order_size": Object {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": Object {
          "amount": "0.00054147",
          "currency": "BTC",
        },
      },
      "ETH": Estimate {
        "coinbase_fees": Object {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": Object {
          "crypto": "ETH",
          "local": "USD",
          "rate": "197.07",
        },
        "exchange_to_proceeds": Object {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": Object {
          "amount": "0.03795",
          "currency": "ETH",
        },
        "fee_per_kb": Object {
          "amount": "0.00086",
          "currency": "ETH",
        },
        "min_order_size": Object {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": Object {
          "amount": "2.53",
          "currency": "ETH",
        },
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "data": Object {
      "estimates": Object {
        "BTC": "BTC",
        "ETH": "ETH",
      },
    },
  },
}
`;
